agent/consul/state/catalog.go

```diff
@@ -922,6 +1033,30 @@ func (s *Store) serviceNodes(ws memdb.WatchSet, serviceName string, connect bool
 		results = append(results, service.(*structs.ServiceNode))
 	}
 
+	// If we are querying for Connect nodes, the associated proxy might be a gateway.
+	// Gateways are tracked in a separate table, and we append them to the result set.
+	if connect {
+		// Look up gateway name associated with the service
+		result, err := s.serviceTerminatingGateway(tx, serviceName, entMeta)
+		if err != nil {
+			return 0, nil, fmt.Errorf("failed gateway lookup: %s", err)
+		}
+		if result != nil {
+			mapping := result.(*structs.GatewayService)
+
+			// Look up nodes for gateway
+			gateways, err := s.catalogServiceNodeList(tx, mapping.Gateway.ID, "service", &mapping.Gateway.EnterpriseMeta)
+			if err != nil {
+				return 0, nil, fmt.Errorf("failed service lookup: %s", err)
+			}
+			ws.Add(gateways.WatchCh())
+
+			for gateway := gateways.Next(); gateway != nil; gateway = gateways.Next() {
+				results = append(results, gateway.(*structs.ServiceNode))
+			}
+		}
```

Comment:
It was intentional to not add it to the watchSet in `checkServiceNodes` because the watches in that function are added for each service name. 

More about there [here](https://github.com/hashicorp/consul/blob/223744e754f2518d779eceef0b89177cb5784057/agent/consul/state/catalog.go#L1963:L1968) and [here](https://github.com/hashicorp/consul/blob/223744e754f2518d779eceef0b89177cb5784057/agent/consul/state/catalog.go#L2013).

Refactored this so that they use a common function.

Reply:
Interesting, did not see those comments before, thanks for pointing them out!

We do still want to add the `gateways.WatchCh()` in this function though, right? That way a blocking query will get notified if a new instance of the terminating gateway is added.

Reply:
Ah yep, missed that in the refactor.

Added a test to keep that in check.